debugger/debugger.js

```diff
@@ -0,0 +1,310 @@
+    this.print("----------------------------------------------------------------------------------\n");
+    this.print(" ==> " + this.get_source_line(lineno - 1) + "\n");
+    this.print("----------------------------------------------------------------------------------\n");
+};
+
+Sk.Debugger.prototype.set_suspension = function(suspension) {
+    var parent = null;
+    if (!hasOwnProperty(suspension, "filename") && suspension.child instanceof Sk.misceval.Suspension) {
+        suspension = suspension.child;
+    }
+        
+    // Pop the last suspension of the stack if there is more than 0
+    if (this.suspension_stack.length > 0) {
+        this.suspension_stack.pop();
+        this.current_suspension -= 1;
+    }
+    
+    // Unroll the stack to get each suspension.
+    while (suspension instanceof Sk.misceval.Suspension) {
+        parent = suspension;
+        this.suspension_stack.push(parent);
+        this.current_suspension += 1;
+        suspension = suspension.child;
+    }
+
+    suspension = parent;
+    
+    this.print_suspension_info(suspension);
+};
+
+Sk.Debugger.prototype.add_breakpoint = function(filename, lineno, colno, temporary) {
+    var key = this.generate_breakpoint_key(filename, lineno, colno);
+    this.dbg_breakpoints[key] = new Sk.Breakpoint(filename, lineno, colno);
+    if (temporary) {
+        this.tmp_breakpoints[key] = true;
+    }
+};
+
+Sk.Debugger.prototype.suspension_handler = function(susp) {
+    return new Promise(function(resolve, reject) {
+        try {
+            resolve(susp.resume());
+        } catch(e) {
+            reject(e);
+        }
+    });
+};
+
+Sk.Debugger.prototype.resume = function() {
+    // Reset the suspension stack to the topmost
+    this.current_suspension = this.suspension_stack.length - 1;
+    
+    if (this.suspension_stack.length === 0) {
+        this.print("No running program");
+    } else {
+        var promise = this.suspension_handler(this.get_active_suspension());
+        promise.then(this.success.bind(this), this.error.bind(this));
+    }
+};
+
+Sk.Debugger.prototype.pop_suspension_stack = function() {
+    this.suspension_stack.pop();
+    this.current_suspension -= 1;
+};
+
+Sk.Debugger.prototype.success = function(r) {
+    if (r instanceof Sk.misceval.Suspension) {
+        this.set_suspension(r);
+    } else {
+        if (this.suspension_stack.length > 0) {
+            // Current suspension needs to be popped of the stack
+            this.pop_suspension_stack();
+            
+            if (this.suspension_stack.length === 0) {
+                this.print("Program execution complete");
+                return;
+            }
+            
+            var parent_suspension = this.get_active_suspension();
+            // The child has completed the execution. So override the child's resume
+            // so we can continue the execution.
+            parent_suspension.child.resume = function() {};
```

Comment:
@rixner I see... that makes sense. I might have misunderstood the whole thing. Let me rework it this way but I think I see your approach and point here.